cryptoki/src/slot/mod.rs

```diff
@@ -163,121 +162,356 @@ impl From<CK_SLOT_INFO> for SlotInfo {
+        self.flags.contains(USER_PIN_TO_BE_CHANGED)
+    }
+
+    /// True if an incorrect Security Officer login PIN has been entered at least once since
+    /// the last successful authentication
+    ///
+    /// **[Conformance](crate#conformance-notes):**
+    /// This value may be set to always be false if the token either does not
+    /// support the functionality or will not reveal the information because of
+    /// its security policy.
+    pub fn so_pin_count_low(&self) -> bool {
+        self.flags.contains(SO_PIN_COUNT_LOW)
     }
 
-    /// Returns the Token's flags
-    pub fn flags(&self) -> TokenFlags {
-        self.val.flags.into()
+    /// True if supplying an incorrect Security Officer PIN will cause it to become locked
+    ///
+    /// **[Conformance](crate#conformance-notes):**
+    /// This value may be set to always be false if the token either does not
+    /// support the functionality or will not reveal the information because of
+    /// its security policy.
+    pub fn so_pin_final_try(&self) -> bool {
+        self.flags.contains(SO_PIN_FINAL_TRY)
+    }
+
+    /// True if the Security Officer PIN has been locked; Security Officer login to the token is not
+    /// possible
+    pub fn so_pin_locked(&self) -> bool {
+        self.flags.contains(SO_PIN_LOCKED)
     }
-}
 
-impl Deref for TokenInfo {
-    type Target = CK_TOKEN_INFO;
+    /// True if the Security Officer PIN value is the default value set by the token
+    /// initialization or manufacturing, or the PIN has been expired by the card
+    ///
+    /// **[Conformance](crate#conformance-notes):**
+    /// This may be always false if the token either does not support the
+    /// functionality or will not reveal the information because of its security
+    /// policy.
+    ///
+    /// If a PIN is set to the default value or has expired, this function
+    /// returns `true`. When true, logging in with a PIN will succeed, but only
+    /// the [`Session::set_pin`][crate::session::Session::set_pin] function can
+    /// be called. Calling any other function that required the user to be
+    /// logged in will cause [`PinExpired`][crate::error::RvError::PinExpired]
+    /// to be returned until
+    /// [`Session::set_pin`][crate::session::Session::set_pin] is called
+    /// successfully.
+    pub fn so_pin_to_be_changed(&self) -> bool {
+        self.flags.contains(SO_PIN_TO_BE_CHANGED)
+    }
 
-    fn deref(&self) -> &Self::Target {
-        &self.val
+    /// True if the token failed a FIPS 140-2 self-test and entered an error state
+    pub fn error_state(&self) -> bool {
+        self.flags.contains(ERROR_STATE)
+    }
+
+    /// The maximum number of sessions that can be opened with the token at one
+    /// time by a single application
+    /// If `None`, this information was unavailable.
+    /// If `Some(None)` there is no maximum, meaning the value is effectively infinite
+    /// If `Some(Some(n))` the maximum number of sessions is `n`
```

Comment:
Would this be cleaner if we used an enum like:

```rust
enum SomethingTokenInfo
{
    Unavailable,
    Infinite,
    Defined(u64)
}
```

Reply:
I definitely considered something like this, and could still be talked into it. I ended up going with this one for a few reasons:
* At the time, I wasn't sure how widely used the two special constants were that make this necessary, and wasn't sold on creating a new type if it turned out to be just this spot. Taking a look at the standard, that does appear to be the case.
* Between the two options, if I later decide I was wrong and want to switch, it's marginally easier to replace standard types with custom ones than the other way around.
* The Option type comes with a lot of functionality attached for free like converting into a Result, and `flatten()` which could be demonstrated in a doc comment.
* Personal bias: This crate (really PCKS#11 itself) requires the use of a lot of jargon like "token" and "slot" and "object" that all have meaning specific to the standard. In that context, I'm a bit biased toward a smaller number of custom types for the sake of having the main ones easier to identify. I wouldn't necessarily feel the same way for other crates.
* Rust provides `Option::None` precisely to be a semantic equivalent to "unavailable" and avoid having to encode it into the type itself (a la a null pointer).

That last item seems to imply that an idomatic new type would look more like
```rust
enum Limit {
    EffectivelyInfinite,
    Defined(u64),
}
```
and then return an `Option<Limit>`. But when I looked at that, I thought, "That's just another`Option` but it does less." and talked myself out of it.